### ISSUE TITLE
New Product Modals - Direct Links to Products

### ISSUE THREAD

**Issue by lin-d-hop** [openfoodfoundation/openfoodnetwork#10879]
When working on this use clokify code #10879 - Product Modal Links

## What is the problem we are solving 

For most e-commerce platforms out there in internet land it is standard practice to have Product Descriptions Pages (PDPs). These pages have direct links and include all the detailed information about a product. For historical reasons within the Open Food Network, we use a type of embedded window called a modal to display this information. Our modals do not have URLs and so it is impossible to link to them.

Direct links to products in the shops are very valuable for marketing reasons, allowing food enterprises to share pictures of specific products within social media feeds or newsletters to help attract people to the online shop. They are also very good for search engine optimisation (SEO), helping to boost the prevalence of the shop in search results when people are searching for food items using a search engine.

This feature would make it easier for people using the Open Food Network software platform to market products sold through the platform. It would mean that they could share direct links to products on their social media channels, as well as on their website, in email and SMS marketing and wherever else was relevant.

Potential benefits include:

- Increasing the amount of interesting content enterprises can generate in their newsletters and social media, by enabling every product in the shop to be a new and interesting potential piece of content. 
- Increasingly the network effect of this content as producers will likely also share links to their own produce. 
- Improving search engine optimisation as each shop will have many more pages that can be scanned and indexed by search engine bots, meaning that shops will land higher in web searches (as will OFN overall). 
- Easier to create campaigns that target specific groups of shoppers based on specific products or sets of products that might be of interest. A great way to target people that aren’t already engaged in your shop.

## List of issues

- [ ] https://github.com/openfoodfoundation/openfoodnetwork/issues/12638

## Success factors = expected outcome 

- A direct URL for a specific product on a shopfront. 
- The ability to click through to where the product is available in a specific order cycle

Jess had a look at this and realised that if we can hide specific elements on this page in the case that this page is displayed after a click on the product modal, then the experience would be much more streamlined for users. Specifically, hiding the Search Bar, the 'Search Results' text and the Filters.

## Previous links

Discussed previously as a potential funded feature, though it's taken a while to get there.
https://openfoodnetwork.org/fundedfeatures2/

Discourse chat:
https://community.openfoodnetwork.org/t/direct-links-to-shop-products-for-marketing-seo-and-more-permalink-deep-link/2176

**Comment by NickIkenma** [comment]
@mariocarabotta, was this covered in the back office UI uplift?

**Comment by lin-d-hop** [comment]
@NickIkenma nope this is awaiting space in our pipe for a kick off

**Comment by mariocarabotta** [comment]
designs are ready for dev here - https://www.figma.com/file/v2tzhoSc7U5pcdurI2tRcD/Engineering-Delivery---Front-Shop?type=design&node-id=1%3A9&mode=design&t=o5hvOukn2AwSCDE7-1

removing design-needed label

**Comment by mariocarabotta** [comment]
flowchart here > https://www.figma.com/board/i9a9fjMjOPy4CiTrJj428h/Product-links?node-id=0-1&t=ungZMexZ2FuRBFEO-1

**Comment by mariocarabotta** [comment]
prototype here > https://www.figma.com/proto/v2tzhoSc7U5pcdurI2tRcD/Engineering-Delivery---Front-Shop?page-id=1%3A9&node-id=1-43&viewport=548%2C492%2C0.1&t=SYSOQYUK0Nbn3Uh5-1&scaling=min-zoom&starting-point-node-id=1%3A43

**Comment by RachL** [comment]
Notes for scenarios / might need dev input:

- when do we have a direct links? E.g. can a shopper land on a product link with an open OC but an out-of-stock product?
- once the shopper wants to see the full store, where do other products of that store appear? => is the product remaining at the top or do we anchor the product at his "usual" place in the shop list of products?
- handle mobile cases: are their cases on mobile we don't want the URL to open in a modal like currently?

**Comment by dacook** [comment]
It looks like we need to work through the requirements a little more before this is ready to develop. In addition to the above queries, I'd like to know:
* The URL format for the product link
* The URL format once you click "Shop for this product"

Also I find that the flow chart is pretty clear, but I find it strange that the modal appears first, which then leads to a full page. I actually thought it should be the other way around.

I'm not sure that Mario has time to discuss this before he goes on leave, so I'm sad to suggest that we park it until we have capacity to refine it further.

**Comment by RachL** [comment]
@dacook yes we went a different way than the flow chart, it was discussed with Mario. Sorry the issue is more precise than the epic : https://github.com/openfoodfoundation/openfoodnetwork/issues/12638

I don't have a recommendation for the URL format. How do we choose one?

**Comment by dacook** [comment]
I would suggest that we use the product name, similar to how we use shop names.  I've implemented this on another project and could make some more specific suggestions when the time comes.

Using the name in the URL has benefits for SEO and human-readability, but it does require some extra management to handle things like when the product gets renamed.
I think we should make them auto-generated, not customisable like enterprise name. This helps avoid mistakes or getting out of date (Eg product "Mixed Veg Box" gets renamed to become "Fruit Box", but they forget to change the url "mixed-veg-box". 
Another thing to handle is when the url changes: ideally we would redirect the old URL to support any old bookmarks, and importantly, search engine results that can take a long time to update. I implemented this by including the product ID in the URL, so that the system can easily recognise it and load the correct product. (There are other options, but this was deemed the least effort)

Product names could go directly under the `shop/` path. But I'd suggest we use a different path just in case we ever want to add other types of pages for shops.
Suggested format: `/shop-name/products/1234-product-name`

To use an example:
* Shop: https://openfoodnetwork.org.au/tcg-fitzroy/shop
* Product name: "Box - $20, Mixed F&V, SMALL"
* Product URL: https://openfoodnetwork.org.au/tcg-fitzroy/products/1234-box-20-mixed-f-and-v-small
* Renamed URL: https://openfoodnetwork.org.au/tcg-fitzroy/products/1234-box-20-fruit-small (the old url would redirect to this one)

 ```ruby
[1234, "Box - $20, Mixed F&V, SMALL".gsub("&", "-and-").parameterize].join("-")
```
(I can share some code with specs from a private project).

**Comment by mkllnk** [comment]
I really like David's solution. It combines the robustness of internal ids and adds the name for UX and SEO. I wonder if we move that logic to a gem?

**Comment by BethanOFN** [comment]
Just adding to the context surrounding this issue, today I had a chat with an ethical marketing firm & freelance food journalist in the UK who are working with one of our producers. They wrote an article recently on the [abysmal ethics of farmed smoked salmon](https://livefrankly.co.uk/sustainable-food/why-wild-smoked-salmon-beats-farmed-every-time-this-christmas/) and wanted to link to our UK fish producer's shopfront, specifically their wild smoked salmon, which they weren't able to do. So as well as helping users with their individual marketing, this feature would also help enable the ethical press to link to specific products

**Comment by RachL** [comment]
alright, I'm trying centralize all notes into one main [issue](https://github.com/openfoodfoundation/openfoodnetwork/issues/12638). closing here
